Track code imports can fail due to deferred dependency install

If an import is not bounded by a try/catch or nested in an invoked function, and the dependency requires #1405, it can fail with an ImportError.
```
[ERROR] Cannot race. Traceback (most recent call last):
  File "/Users/rick.boyd/scratch/test_geneve/.venv/lib/python3.9/site-packages/esrally/utils/modules.py", line 106, in load
    root_module = self._load_component(component_name, module_dirs)
  File "/Users/rick.boyd/scratch/test_geneve/.venv/lib/python3.9/site-packages/esrally/utils/modules.py", line 65, in _load_component
    m = importlib.import_module(p)
  File "/Users/rick.boyd/.pyenv/versions/3.9.0/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 790, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/Users/rick.boyd/.rally/benchmarks/tracks/cavokz/solutions/security/track.py", line 30, in <module>
    from security.parameter_sources.events_emitter import EventsEmitterParamSource
  File "/Users/rick.boyd/.rally/benchmarks/tracks/cavokz/solutions/security/parameter_sources/events_emitter.py", line 22, in <module>
    import geneve
ModuleNotFoundError: No module named 'geneve'
```

Likewise, but more flaky, I've seen `used_corpora` call invoke the param source function where we use a dependency but fail to resolve it in the PYTHONPATH in time, via the following stack:
```
esrally.exceptions.RallyError: Traceback (most recent call last):
  File "/Users/rick.boyd/scratch/test_geneve/.venv/lib/python3.9/site-packages/esrally/actor.py", line 92, in guard
    return f(self, msg, sender)
  File "/Users/rick.boyd/scratch/test_geneve/.venv/lib/python3.9/site-packages/esrally/driver/driver.py", line 535, in receiveMsg_WorkerIdle
    self.transition_when_all_children_responded(sender, msg, self.Status.PROCESSOR_RUNNING, self.Status.PROCESSOR_COMPLETE, self.resume)
  File "/Users/rick.boyd/scratch/test_geneve/.venv/lib/python3.9/site-packages/esrally/actor.py", line 157, in transition_when_all_children_responded
    transition()
  File "/Users/rick.boyd/scratch/test_geneve/.venv/lib/python3.9/site-packages/esrally/driver/driver.py", line 510, in resume
    self._seed_tasks(self.processors.get())
  File "/Users/rick.boyd/scratch/test_geneve/.venv/lib/python3.9/site-packages/esrally/driver/driver.py", line 518, in _seed_tasks
    self.tasks = list(WorkerTask(func, params) for func, params in processor.on_prepare_track(self.track, self.data_root_dir))
  File "/Users/rick.boyd/scratch/test_geneve/.venv/lib/python3.9/site-packages/esrally/driver/driver.py", line 518, in <genexpr>
    self.tasks = list(WorkerTask(func, params) for func, params in processor.on_prepare_track(self.track, self.data_root_dir))
  File "/Users/rick.boyd/scratch/test_geneve/.venv/lib/python3.9/site-packages/esrally/track/loader.py", line 467, in on_prepare_track
    for corpus in used_corpora(track):
  File "/Users/rick.boyd/scratch/test_geneve/.venv/lib/python3.9/site-packages/esrally/track/loader.py", line 434, in used_corpora
    param_source = operation_parameters(t, sub_task)
  File "/Users/rick.boyd/scratch/test_geneve/.venv/lib/python3.9/site-packages/esrally/track/loader.py", line 423, in operation_parameters
    return params.param_source_for_name(op.param_source, t, op.params)
  File "/Users/rick.boyd/scratch/test_geneve/.venv/lib/python3.9/site-packages/esrally/track/params.py", line 51, in param_source_for_name
    return param_source(track, params)
  File "/Users/rick.boyd/.rally/benchmarks/tracks/cavokz/solutions/security/parameter_sources/events_emitter.py", line 84, in __init__
    schema = kwargs["_test_schema"] if "_test_schema" in kwargs else load_schema(track, params)
  File "/Users/rick.boyd/.rally/benchmarks/tracks/cavokz/solutions/security/parameter_sources/events_emitter.py", line 44, in load_schema
    return geneve.utils.load_schema(params["schema"]["uri"], params["schema"]["path"], track.root)
NameError: name 'geneve' is not defined
```

The second error seems to only happen the first time a track is run. We should (a) ensure dependency installation happens as early as possible and (b) (stretch goal) get `param_source.__init__()` out of the path of execution for the track preparator, if possible.

I will attach a simple reproduction soon

--- a/esrally/driver/driver.py
+++ b/esrally/driver/driver.py
@@ -473,7 +473,7 @@ class TrackPreparationActor(actor.RallyActor):
         # load node-specific config to have correct paths available
         self.cfg = load_local_config(msg.config)
         # this instance of load_track occurs once per host, so install dependencies if necessary
-        load_track(self.cfg, install_dependencies=True)
+        load_track(self.cfg, install_dependencies=False)
         self.send(sender, ReadyForWork())
 
     @actor.no_retry("track preparator")  # pylint: disable=no-value-for-parameter

--- a/esrally/racecontrol.py
+++ b/esrally/racecontrol.py
@@ -191,7 +191,7 @@ class BenchmarkCoordinator:
             if specified_version < min_es_version:
                 raise exceptions.SystemSetupError(f"Cluster version must be at least [{min_es_version}] but was [{distribution_version}]")
 
-        self.current_track = track.load_track(self.cfg, install_dependencies=False)
+        self.current_track = track.load_track(self.cfg, install_dependencies=True)
         self.track_revision = self.cfg.opts("track", "repository.revision", mandatory=False)
         challenge_name = self.cfg.opts("track", "challenge.name")
         self.current_challenge = self.current_track.find_challenge_or_default(challenge_name)

--- a/esrally/rally.py
+++ b/esrally/rally.py
@@ -20,6 +20,7 @@ import datetime
 import logging
 import os
 import platform
+import shutil
 import sys
 import time
 import uuid
@@ -1068,6 +1069,17 @@ def main():
             console.warn("No Internet connection detected. Specify --offline to run without it.", logger=logger)
             sys.exit(0)
         logger.info("Detected a working Internet connection.")
+
+    def _trap(function, path, exc_info):
+        if exc_info[0] == FileNotFoundError:
+            # couldn't delete because it was already clean
+            return
+        logging.exception("Failed to clean up [%s] with [%s]", path, function, exc_info=True)
+        raise exceptions.SystemSetupError(f"Unable to clean [{paths.libs()}]. See Rally log for more information.")
+
+    # fully destructive is fine, we only allow one Rally to run at a time and we will rely on the pip cache for download caching
+    logging.info("Cleaning track dependency directory [%s]...", paths.libs())
+    shutil.rmtree(paths.libs(), onerror=_trap)
 
     result = dispatch_sub_command(arg_parser, args, cfg)
 

--- a/esrally/track/loader.py
+++ b/esrally/track/loader.py
@@ -195,21 +195,7 @@ def load_track(cfg, install_dependencies=False):
 
 
 def _install_dependencies(dependencies):
-    def _cleanup():
-        # fully destructive is fine, we only allow one Rally to run at a time and we will rely on the pip cache for download caching
-        console.info("Cleaning track dependency directory...")
-        logging.info("Cleaning track dependency directory [%s]...", paths.libs())
-        shutil.rmtree(paths.libs(), onerror=_trap)
-
-    def _trap(function, path, exc_info):
-        if exc_info[0] == FileNotFoundError:
-            # couldn't delete because it was already clean
-            return
-        logging.exception("Failed to clean up [%s] with [%s]", path, function, exc_info=True)
-        raise exceptions.SystemSetupError(f"Unable to clean [{paths.libs()}]. See Rally log for more information.")
-
     if dependencies:
-        _cleanup()
         log_path = os.path.join(paths.logs(), "dependency.log")
         console.info(f"Installing track dependencies [{', '.join(dependencies)}]")
         try:

--- a/esrally/track/loader.py
+++ b/esrally/track/loader.py
@@ -228,7 +228,9 @@ def _load_single_track(cfg, track_repository, track_name, install_dependencies=F
         tpr = TrackProcessorRegistry(cfg)
         if install_dependencies:
             _install_dependencies(current_track.dependencies)
-        has_plugins = load_track_plugins(cfg, track_name, register_track_processor=tpr.register_track_processor)
+        has_plugins = load_track_plugins(
+            cfg, track_name, install_dependencies=install_dependencies, register_track_processor=tpr.register_track_processor
+        )
         current_track.has_plugins = has_plugins
         for processor in tpr.processors:
             processor.on_after_load_track(current_track)
@@ -271,7 +273,7 @@ def load_track_plugins(
     plugin_reader = TrackPluginReader(track_plugin_path, register_runner, register_scheduler, register_track_processor)
 
     if plugin_reader.can_load():
-        plugin_reader.load()
+        plugin_reader.load(install_dependencies=install_dependencies)
         return True
     else:
         return False
@@ -1121,10 +1123,11 @@ class TrackPluginReader:
     def can_load(self):
         return self.loader.can_load()
 
-    def load(self):
-        # get dependent libraries installed in a prior step. ensure dir exists to make sure loading works correctly.
-        os.makedirs(paths.libs(), exist_ok=True)
-        sys.path.insert(0, paths.libs())
+    def load(self, install_dependencies):
+        if install_dependencies:
+            # get dependent libraries installed in a prior step. ensure dir exists to make sure loading works correctly.
+            os.makedirs(paths.libs(), exist_ok=True)
+            sys.path.insert(0, paths.libs())
         root_module = self.loader.load()
         try:
             # every module needs to have a register() method

--- a/esrally/track/loader.py
+++ b/esrally/track/loader.py
@@ -202,6 +202,9 @@ def _install_dependencies(dependencies):
         shutil.rmtree(paths.libs(), onerror=_trap)
 
     def _trap(function, path, exc_info):
+        if exc_info[0].__name__ == FileNotFoundError.__name__:
+            # couldn't delete because it was already clean
+            return
         logging.exception("Failed to clean up [%s] with [%s]", path, function, exc_info=True)
         raise exceptions.SystemSetupError(f"Unable to clean [{paths.libs()}]. See Rally log for more information.")
 

--- a/esrally/track/loader.py
+++ b/esrally/track/loader.py
@@ -19,7 +19,6 @@ import json
 import logging
 import os
 import re
-import shutil
 import subprocess
 import sys
 import tempfile
@@ -238,7 +237,6 @@ def load_track_plugins(
     register_runner=None,
     register_scheduler=None,
     register_track_processor=None,
-    install_dependencies=False,
     force_update=False,
 ):
     """
@@ -249,7 +247,6 @@ def load_track_plugins(
     :param register_runner: An optional function where runners can be registered.
     :param register_scheduler: An optional function where custom schedulers can be registered.
     :param register_track_processor: An optional function where track processors can be registered.
-    :param install_dependencies: If set to ``True``, install declared dependencies from the track.py. Defaults to ``False``.
     :param force_update: If set to ``True`` this ensures that the track is first updated from the remote repository.
                          Defaults to ``False``.
     :return: True iff this track defines plugins and they have been loaded.

--- a/esrally/track/loader.py
+++ b/esrally/track/loader.py
@@ -231,9 +231,7 @@ def _load_single_track(cfg, track_repository, track_name, install_dependencies=F
         tpr = TrackProcessorRegistry(cfg)
         if install_dependencies:
             _install_dependencies(current_track.dependencies)
-        has_plugins = load_track_plugins(
-            cfg, track_name, install_dependencies=install_dependencies, register_track_processor=tpr.register_track_processor
-        )
+        has_plugins = load_track_plugins(cfg, track_name, register_track_processor=tpr.register_track_processor)
         current_track.has_plugins = has_plugins
         for processor in tpr.processors:
             processor.on_after_load_track(current_track)
@@ -276,7 +274,7 @@ def load_track_plugins(
     plugin_reader = TrackPluginReader(track_plugin_path, register_runner, register_scheduler, register_track_processor)
 
     if plugin_reader.can_load():
-        plugin_reader.load(install_dependencies=install_dependencies)
+        plugin_reader.load()
         return True
     else:
         return False
@@ -1126,11 +1124,10 @@ class TrackPluginReader:
     def can_load(self):
         return self.loader.can_load()
 
-    def load(self, install_dependencies):
-        if install_dependencies:
-            # get dependent libraries installed in a prior step. ensure dir exists to make sure loading works correctly.
-            os.makedirs(paths.libs(), exist_ok=True)
-            sys.path.insert(0, paths.libs())
+    def load(self):
+        # get dependent libraries installed in a prior step. ensure dir exists to make sure loading works correctly.
+        os.makedirs(paths.libs(), exist_ok=True)
+        sys.path.insert(0, paths.libs())
         root_module = self.loader.load()
         try:
             # every module needs to have a register() method

--- a/esrally/track/loader.py
+++ b/esrally/track/loader.py
@@ -202,7 +202,7 @@ def _install_dependencies(dependencies):
         shutil.rmtree(paths.libs(), onerror=_trap)
 
     def _trap(function, path, exc_info):
-        if exc_info[0].__name__ == FileNotFoundError.__name__:
+        if exc_info[0] == FileNotFoundError:
             # couldn't delete because it was already clean
             return
         logging.exception("Failed to clean up [%s] with [%s]", path, function, exc_info=True)

